staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+	corev1.Resource("pods"):                   true,
+	corev1.Resource("replicationcontrollers"): true,
+	corev1.Resource("podtemplates"):           true,
+	appsv1.Resource("replicasets"):            true,
+	appsv1.Resource("deployments"):            true,
+	appsv1.Resource("statefulsets"):           true,
+	appsv1.Resource("daemonsets"):             true,
+	batchv1.Resource("jobs"):                  true,
+	batchv1.Resource("cronjobs"):              true,
+}
+
+type DefaultPodSpecExtractor struct{}
+
+func (DefaultPodSpecExtractor) HasPodSpec(gr schema.GroupResource) bool {
+	return defaultPodSpecResources[gr]
+}
+
+func (DefaultPodSpecExtractor) ExtractPodSpec(obj runtime.Object) (*metav1.ObjectMeta, *corev1.PodSpec, error) {
+	switch o := obj.(type) {
+	case *corev1.Pod:
+		return &o.ObjectMeta, &o.Spec, nil
+	case *corev1.PodTemplate:
+		return extractPodSpecFromTemplate(&o.Template)
+	case *corev1.ReplicationController:
+		return extractPodSpecFromTemplate(o.Spec.Template)
+	case *appsv1.ReplicaSet:
+		return extractPodSpecFromTemplate(&o.Spec.Template)
+	case *appsv1.Deployment:
+		return extractPodSpecFromTemplate(&o.Spec.Template)
+	case *appsv1.DaemonSet:
+		return extractPodSpecFromTemplate(&o.Spec.Template)
+	case *appsv1.StatefulSet:
+		return extractPodSpecFromTemplate(&o.Spec.Template)
+	case *batchv1.Job:
+		return extractPodSpecFromTemplate(&o.Spec.Template)
+	case *batchv1.CronJob:
+		return extractPodSpecFromTemplate(&o.Spec.JobTemplate.Spec.Template)
+	default:
+		return nil, nil, fmt.Errorf("unexpected object type: %s", obj.GetObjectKind().GroupVersionKind().String())
+	}
+}
+
+func (DefaultPodSpecExtractor) PodSpecResources() []schema.GroupResource {
+	retval := make([]schema.GroupResource, 0, len(defaultPodSpecResources))
+	for r := range defaultPodSpecResources {
+		retval = append(retval, r)
+	}
+	return retval
+}
+
+func extractPodSpecFromTemplate(template *corev1.PodTemplateSpec) (*metav1.ObjectMeta, *corev1.PodSpec, error) {
+	if template == nil {
+		return nil, nil, nil
+	}
+	return &template.ObjectMeta, &template.Spec, nil
+}
+
+// CompleteConfiguration() sets up default or derived configuration.
+func (a *Admission) CompleteConfiguration() error {
+	if a.Configuration != nil {
+		if p, err := admissionapi.ToPolicy(a.Configuration.Defaults); err != nil {
+			return err
+		} else {
+			a.defaultPolicy = p
+		}
+	}
+
+	if a.PodSpecExtractor == nil {
+		a.PodSpecExtractor = &DefaultPodSpecExtractor{}
+	}
+
+	return nil
+}
+
+// ValidateConfiguration() ensures all required fields are set with valid values.
+func (a *Admission) ValidateConfiguration() error {
+	if a.Configuration == nil {
+		return fmt.Errorf("configuration required")
+	} else if errs := validation.ValidatePodSecurityConfiguration(a.Configuration); len(errs) > 0 {
+		return errs.ToAggregate()
+	} else {
+		if p, err := admissionapi.ToPolicy(a.Configuration.Defaults); err != nil {
+			return err
+		} else if !reflect.DeepEqual(p, a.defaultPolicy) {
```

Comment:
note to self, track this down.

Reply:
will clean up, this is trying to make sure Complete() was run and the results match (to avoid mutating in validation)

Reply:
after frowning at this a bit, I still don't want to mutate in validation, or after calling validation... this feels awkward, but I might leave it for now (externally, calling Complete, then Validate feels ok)